### ISSUE TITLE
DTLSv1 client fails if server sends certificate request

### ISSUE THREAD

**Issue by umokk** [libressl-portable/portable#291]
Running your own tests/ssltest.c in "test dtlsv1 with client authentication" mode gives "ERROR IN STARTUP" or "ERROR in SERVER". The problem is server doesn't get client's certificate. The same test works fine with OpenSSL.

**Comment by kinichiro** [comment]
I think this 'break' in ssltest.c should be 'goto err' or set 'ret' to 0 before 'break' to catch this error.

```
...
   1400                         if (!do_client && !do_server) {
   1401                         fprintf(stdout, "ERROR IN STARTUP\n");
   1402                         ERR_print_errors(bio_err);
   1403                         break;
...
```

It appears dtls1 server function has some issue.
I tested this issue with openssl(1) s_server and s_client command, too.
I used OpenSSL1.0.2 and LibreSSL2.5.2(current) on Fedora.
With s_server = OpenSSL, both s_client succeeded to connect.
With s_server = LibreSSL, both s_client failed with errno=111.

Log s_server = LibreSSL was like this.

```
$ ../apps/openssl/openssl s_server -dtls1 -cert server.pem -msg
Using auto DH parameters
Using default temp ECDH parameters
ACCEPT
```

Log s_client was like this.

```
$ openssl s_client -dtls1 -port 4433 -msg
CONNECTED(00000003)
>>> ??? [length 000d]
    16 fe ff 00 00 00 00 00 00 00 00 00 81
>>> DTLS 1.0 Handshake [length 0081], ClientHello
    01 00 00 75 00 00 00 00 00 00 00 75 fe ff d2 76
    f6 3e 01 2c d6 e2 73 37 ca 75 3f 51 3f b8 63 c0
    71 77 4b 73 15 e1 e4 b5 42 a5 a3 69 3b 8d 00 00
    00 2c c0 14 c0 0a c0 13 c0 09 00 35 00 84 00 2f
    00 41 00 39 00 38 00 88 00 87 00 33 00 32 00 45
    00 44 c0 12 c0 08 00 0a 00 16 00 13 00 ff 01 00
    00 1f 00 0b 00 04 03 00 01 02 00 0a 00 0a 00 08
    00 17 00 19 00 18 00 16 00 23 00 00 00 0f 00 01
    01
write:errno=111
---
no peer certificate available
---
No client certificate CA names sent
---
SSL handshake has read 0 bytes and written 0 bytes
---
New, (NONE), Cipher is (NONE)
Secure Renegotiation IS NOT supported
Compression: NONE
Expansion: NONE
No ALPN negotiated
SSL-Session:
    Protocol  : DTLSv1
    Cipher    : 0000
    Session-ID:
    Session-ID-ctx:
    Master-Key:
    Key-Arg   : None
    Krb5 Principal: None
    PSK identity: None
    PSK identity hint: None
    Start Time: 1488611346
    Timeout   : 7200 (sec)
    Verify return code: 0 (ok)
---
```

2.4.5 has same issue.

**Comment by 4a6f656c** [comment]
@umokk thanks for the report - this is now fixed in r1.8 of https://github.com/openbsd/src/commits/master/lib/libssl/ssl_clnt.c.

@kinichiro would you mind retesting and then closing this issue?

**Comment by umokk** [comment]
@kinichiro 
Changing **break** to **goto err** just makes ssltest skip all further tests. Nothing useful :(

@4a6f656c 
Give me some time to test it :)

**Comment by kinichiro** [comment]
@umokk Original issue was fixed by https://github.com/libressl-portable/openbsd/commit/3f7ddbfb2608b4b539f2a5e7660c9658eca398d8 .
For ssltest.c, this is regression test and exits soon when it finds error,
as same as other regression tests in LibreSSL.
Thanks for informing this issue :)

@4a6f656c openssl(1) s_server and s_client DTLS connection issue still remains.
But, this issue doesn't occur on OpenBSD 6.0, but on Linux (I saw on Fedora23).
It appears to be a platform specific issue, and I think it is differ from original issue.
It should be discussed by another topic, and I will close this issue.

**Comment by kinichiro** [comment]
I found s_client needed explicitly host arguments on my Linux environment.
The error message I encountered `write:errno=111` was ECONNREFUSED.
This way of executing s_client succeeded.

```
$ openssl s_client -dtls1 -host 127.0.0.1 -port 4433 -msg
```

The /etc/hosts file on my Linux had only loopback address and "localhost",
didn't have my computer name.

```
127.0.0.1   localhost
::1         localhost
```

I would not create another issue for this, and sorry for my noise.
